File: src/test/baekjoon/python/dfsbfs/Tree.py
n = int(input())
tree = list(map(int, input().split()))
k = int(input())
graph = [[] for _ in range(n)]
answer = 0
root = -1

def dfs(node):
    global answer
    if not graph[node]:
        answer += 1
        return

    children = graph[node]
    for i in range(len(children)):
        if children[i] == k:
            if len(children) == 1:
                answer += 1
            else:
                continue
        else:
            dfs(children[i])


for i in range(n):
    if tree[i] == -1:
        root = i
    else:
        graph[tree[i]].append(i)


if k == root:
    print(0)
else:
    dfs(root)
    print(answer)

File: src/test/baekjoon/java/dfsbfs/Tree.java
package test.baekjoon.java.dfsbfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Tree {

    static List<List<Integer>> graph;
    static int answer = 0;
    static int k;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        k = Integer.parseInt(br.readLine());
        graph = new ArrayList<>();
        int[] tree = new int[n];
        int root = -1;

        for (int i = 0; i < n; i++) {
            tree[i] = Integer.parseInt(st.nextToken());
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < n; i++) {
            if (tree[i] == -1) {
                root = i;
            } else {
                graph.get(tree[i]).add(i);
            }
        }

        if (root == k) {
            System.out.println(0);
        } else {
            dfs(root);
            System.out.println(answer);
        }
    }

    static void dfs(int node) {
        if (graph.get(node).isEmpty()) {
            answer++;

            return;
        }

        List<Integer> children = graph.get(node);

        for (int i = 0; i < children.size(); i++) {
            if (children.get(i) == k) {
                if (children.size() == 1) {
                    answer++;
                }
            } else {
                dfs(children.get(i));
            }
        }
    }
}
